metagraph/src/cli/align.cpp

```diff
@@ -397,68 +407,95 @@ int align_to_graph(Config *config) {
 
     for (const auto &file : files) {
         logger->trace("Align sequences from file '{}'", file);
+        seq_io::FastaParser fasta_parser(file, config->forward_and_reverse);
 
         Timer data_reading_timer;
 
         std::ostream *out = config->outfbase.size()
             ? new std::ofstream(config->outfbase)
             : &std::cout;
 
-        seq_io::read_fasta_file_critical(file, [&](kseq_t *read_stream) {
-            thread_pool.enqueue([&](const std::string &query, const std::string &header) {
-                auto paths = aligner->align(query);
-
-                std::lock_guard<std::mutex> lock(print_mutex);
-                if (!config->output_json) {
-                    *out << header << "\t" << paths.get_query();
-                    if (paths.empty()) {
-                        *out << "\t*\t*\t" << config->alignment_min_path_score
-                             << "\t*\t*\t*";
-                    } else {
-                        for (const auto &path : paths) {
-                            *out << "\t" << path;
-                        }
-                    }
-
-                    *out << "\n";
-                } else {
-                    Json::StreamWriterBuilder builder;
-                    builder["indentation"] = "";
-
-                    bool secondary = false;
-                    for (const auto &path : paths) {
-                        const auto& path_query = path.get_orientation()
-                            ? paths.get_query_reverse_complement()
-                            : paths.get_query();
-
-                        *out << Json::writeString(builder,
-                                                  path.to_json(path_query,
-                                                               *graph,
-                                                               secondary,
-                                                               header)) << "\n";
+        const uint64_t batch_size = config->query_batch_size_in_bytes;
+
+        auto it = fasta_parser.begin();
+        auto end = fasta_parser.end();
+        while (it != end) {
+            uint64_t num_bytes_read = 0;
+            // A generator that can be called multiple times until all sequences
+            // are called.
+            std::vector<std::pair<std::string, std::string>> seq_batch;
+            num_bytes_read = 0;
+            for ( ; it != end && num_bytes_read <= batch_size; ++it) {
+                std::string header
+                    = config->fasta_anno_comment_delim != Config::UNINITIALIZED_STR
+                        && it->comment.l
+                            ? utils::join_strings({ it->name.s, it->comment.s },
+                                                  config->fasta_anno_comment_delim,
+                                                  true)
+                            : std::string(it->name.s);
```

Comment:
```suggestion
                std::string header
                    = config->fasta_anno_comment_delim != Config::UNINITIALIZED_STR
                            ? utils::join_strings({ it->name.s, it->comment.s },
                                                  config->fasta_anno_comment_delim,
                                                  true)
                            : std::string(it->name.s);
```

Reply:
this should be equivalent

Reply:
I tried this and it throws an exception because `it->comment.s == NULL`

Reply:
sounds like a bug that needs to be fixed

Reply:
ah, no, no bug